### PR TITLE
Cache keys

### PR DESCRIPTION
**Pull request by cammellos**
Generating a symkey can take up to a second on slow devices, this commit
makes so that keys are saved once generated and stored in the database.

### CODE REVIEW

File: static/chat_db_migrations/1561059285_add_whisper_keys.up.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE whisper_keys (
+  chat_id TEXT NOT NULL PRIMARY KEY ON CONFLICT IGNORE,
```
Comment (dshulyak) [inline static/chat_db_migrations/1561059285_add_whisper_keys.up.sql]: nit: not null is not necessary with primary key

File: static/chat_db_migrations/1561059285_add_whisper_keys.up.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+  key BLOB NOT NULL
+)
```
Comment (dshulyak) [inline static/chat_db_migrations/1561059285_add_whisper_keys.up.sql]: without rowid can be used

File: services/shhext/filter/service.go [hunk 3]
```diff
@@ -390,19 +401,36 @@ func (s *Service) loadContactCode(identity string) (*Chat, error) {
...
 	topics := [][]byte{topic}
 
-	symKeyID, err := s.whisper.AddSymKeyFromPassword(chatID)
-	if err != nil {
-		log.Error("SYM KEYN FAILED", "err", err)
-		return nil, err
-	}
+	symKey, ok := s.keys[chatID]
+	if ok {
+		log.Debug("Loading from cache")
```
Comment (dshulyak) [inline services/shhext/filter/service.go]: maybe add a chat id to log?

**Review by dshulyak — APPROVED** [review]

File: static/chat_db_migrations/1561059285_add_whisper_keys.up.sql (hunk 2 above)
Comment (dshulyak) [reply static/chat_db_migrations/1561059285_add_whisper_keys.up.sql]: @cammellos any reason not to? not trying to enforce, just curious why

Comment (cammellos) [reply static/chat_db_migrations/1561059285_add_whisper_keys.up.sql]: no reason, didn't think of it, but we don't need `rowid` so I have added it now, thanks!

File: services/shhext/filter/persistence.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+type SQLLitePersistence struct {
+	db *sql.DB
+}
+
+func NewSQLLitePersistence(db *sql.DB) *SQLLitePersistence {
+	return &SQLLitePersistence{db: db}
+}
+
+func (s *SQLLitePersistence) All() (map[string][]byte, error) {
+
```
Comment (adambabik) [inline services/shhext/filter/persistence.go]:
```suggestion
```

File: services/shhext/filter/persistence.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+			return nil, err
+		}
+		keys[chatID] = key
+	}
+
+	return keys, nil
+}
+
+func (s *SQLLitePersistence) Add(chatID string, key []byte) error {
+	statement := "INSERT INTO whisper_keys(chat_id,key) VALUES(?,?)"
```
Comment (adambabik) [inline services/shhext/filter/persistence.go]:
```suggestion
	statement := "INSERT INTO whisper_keys(chat_id, key) VALUES(?, ?)"
```

**Review by adambabik — APPROVED** [review]

**Pull request merged**